<?xml version="1.0"?>
<launch>
    <arg name="tfl_detector"        default="none"      doc="carla, mqtt, camera, mqtt_camera_fusion" />
    <arg name="lanelet2_map_name"   default="tartu_demo.osm"    doc="for carla traffic light detector" />
    <arg name="onnx_path"           default="$(find autoware_mini)/config/traffic_lights/tlr_model.onnx" doc="onnx model path" />

    <!-- Carla's ground truth traffic lights -->
    <node if="$(eval tfl_detector=='carla')" pkg="autoware_mini" type="carla_traffic_light_detector.py" name="carla_traffic_light_detector" output="screen" required="true">
        <param name="lanelet2_map_name" value="$(find autoware_mini)/data/maps/$(arg lanelet2_map_name)" />
    </node>

    <!-- MQTT traffic light detector -->
    <node if="$(eval 'mqtt' in tfl_detector)" pkg="autoware_mini" type="mqtt_traffic_light_detector.py" name="mqtt_traffic_light_detector" output="screen" required="true">
        <param name="lanelet2_map_name" value="$(find autoware_mini)/data/maps/$(arg lanelet2_map_name)" />
        <remap if="$(eval tfl_detector=='mqtt_camera_fusion')" from="traffic_light_status" to="mqtt/traffic_light_status"/>
    </node>

    <!-- Camera based traffic light detection -->
    <group if="$(eval 'camera' in tfl_detector)">
        
        <!-- Camera FL -->
        <node pkg="autoware_mini" type="camera_traffic_light_detector.py" name="camera_traffic_light_detector_left" output="screen" required="true" ns="camera_fl">
            <remap from="camera_info" to="/camera_fl/camera_info"/>
            <remap from="image_raw" to="/camera_fl/image_raw"/>
            <param name="lanelet2_map_name" value="$(find autoware_mini)/data/maps/$(arg lanelet2_map_name)" />
            <param name="onnx_path" value="$(arg onnx_path)" />
        </node>

        <!-- Camera FR -->
        <node pkg="autoware_mini" type="camera_traffic_light_detector.py" name="camera_traffic_light_detector_right" output="screen" required="true" ns="camera_fr">
            <remap from="camera_info" to="/camera_fr/camera_info"/>
            <remap from="image_raw" to="/camera_fr/image_raw"/>
            <param name="lanelet2_map_name" value="$(find autoware_mini)/data/maps/$(arg lanelet2_map_name)" />
            <param name="onnx_path" value="$(arg onnx_path)" />
        </node>

        <node pkg="autoware_mini" type="traffic_light_majority_merger.py" name="traffic_light_majority_merger" output="screen" required="true" >
            <remap if="$(eval tfl_detector=='mqtt_camera_fusion')" from="traffic_light_status" to="camera/traffic_light_status"/>
        </node>

        <node if="$(eval tfl_detector=='mqtt_camera_fusion')" pkg="autoware_mini" type="traffic_light_priority_merger.py" name="traffic_light_priority_merger" output="screen" required="true" >
            <remap from="tfl_status_topic_1" to="mqtt/traffic_light_status"/>
            <remap from="tfl_status_topic_2" to="camera/traffic_light_status"/>
        </node>

    </group>

</launch>